File: tests/e2e/harness/tasks.py
"""Background-task helpers.

Every mutating admin action (create site, install/uninstall app, drop site) runs
as a forked background task. The UI fires the request, gets a task_id, and
streams progress. For tests we capture that task_id from the response and then
poll the task API to completion — far more robust than racing UI toasts.
"""

from __future__ import annotations

import time
from typing import Callable

from playwright.sync_api import APIRequestContext, Page, expect


def run_task_action(
    page: Page,
    url_fragment: str,
    action: Callable[[], None],
    method: str = "POST",
) -> str:
    """Run a UI action that kicks off a background task and return its task_id.

    Pass the URL fragment the action POSTs to (e.g. 'create', 'install-app',
    'drop') so we wait for the right response.
    """
    with page.expect_response(
        lambda r: url_fragment in r.url and r.request.method == method
    ) as response_info:
        action()
    body = response_info.value.json()
    if not body.get("task_id"):
        raise RuntimeError(f'Action "{url_fragment}" did not start a task: {body}')
    return body["task_id"]


def wait_for_task(
    request: APIRequestContext,
    base_url: str,
    task_id: str,
    timeout: float = 30 * 60,
) -> None:
    """Poll /api/v1/tasks/:id until it succeeds; raise with the output tail on failure."""
    deadline = time.time() + timeout
    while time.time() < deadline:
        res = request.get(f"{base_url}/api/v1/tasks/{task_id}")
        if res.ok:
            data = res.json()
            status = data["status"]
            if status == "success":
                return
            if status == "failed":
                output = request.get(
                    f"{base_url}/api/v1/tasks/{task_id}/output/content"
                )
                tail = "\n".join(output.text().splitlines()[-30:]) if output.ok else ""
                raise RuntimeError(f"Task {task_id} failed:\n{tail}")
        time.sleep(2)
    raise TimeoutError(f"Task {task_id} did not finish within {timeout}s")


def run_and_await_task(
    page: Page,
    base_url: str,
    url_fragment: str,
    action: Callable[[], None],
    timeout: float = 30 * 60,
) -> None:
    """Convenience: run a task-producing action and wait for it to succeed."""
    task_id = run_task_action(page, url_fragment, action)
    wait_for_task(page.request, base_url, task_id, timeout)


def expect_bench_online(request: APIRequestContext, base_url: str) -> None:
    """Sanity assert that the admin is reachable and out of wizard mode."""
    res = request.get(f"{base_url}/api/v1/bootstrap")
    expect(res).to_be_ok()
    assert res.json().get("mode") == "admin"

File: admin/frontend/src/api/tasks.js
import { apiUrl, request } from './client'

export const tasksApi = {
  list: (status) =>
    request.get('tasks', status && status !== 'all' ? { searchParams: { status } } : {}).json(),
  detail: (taskId) => request.get(`tasks/${taskId}`).json(),
  run: (command, args = {}) => request.post('tasks', { json: { command, ...args } }).json(),
  cancel: (taskId) => request.delete(`tasks/${taskId}`),
  retry: (taskId) => request.post(`tasks/${taskId}/actions/retry`).json(),
  output: async (taskId) => {
    const response = await request.get(`tasks/${taskId}/output/content`)
    return response.ok ? response.text() : ''
  },
  outputUrl: (taskId) => apiUrl(`tasks/${taskId}/output/content`),
  streamUrl: (taskId) => apiUrl(`tasks/${taskId}/events`),
  debugUrl: (taskId) => apiUrl(`tasks/${taskId}/debug`),
}

export const taskWorkerApi = {
  detail: () => request.get('task-worker').json(),
  start: () => request.post('task-worker/actions/start').json(),
  stop: () => request.post('task-worker/actions/stop').json(),
}
